File: sample/src/main/java/io/piano/sample/SimpleDependenciesProvider.java
package io.piano.sample;

import android.content.Context;

import androidx.annotation.NonNull;

import com.squareup.moshi.Moshi;

import io.piano.android.id.PianoIdTokenJsonAdapter;

public class SimpleDependenciesProvider {
    private static SimpleDependenciesProvider instance;

    private final Context appContext;
    private final Moshi moshi;
    private final PrefsStorage prefsStorage;

    private SimpleDependenciesProvider(Context context) {
        appContext = context.getApplicationContext();
        moshi = new Moshi.Builder().add(PianoIdTokenJsonAdapter.FACTORY).build();
        prefsStorage = new PrefsStorage(appContext, moshi);
    }

    public Context getAppContext() {
        return appContext;
    }

    public Moshi getMoshi() {
        return moshi;
    }

    public PrefsStorage getPrefsStorage() {
        return prefsStorage;
    }

    @NonNull
    public static SimpleDependenciesProvider getInstance(Context context) {
        if (instance == null) {
            synchronized (SimpleDependenciesProvider.class) {
                if (instance == null)
                    instance = new SimpleDependenciesProvider(context);
            }
        }
        return instance;
    }
}

File: sample/src/main/java/io/piano/sample/SimpleDependenciesProvider.kt
package io.piano.sample

import android.content.Context
import com.squareup.moshi.Moshi
import io.piano.android.id.PianoIdJsonAdapterFactory
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor

class SimpleDependenciesProvider private constructor(
    context: Context,
) {
    val appContext: Context = context.applicationContext
    val moshi: Moshi = Moshi.Builder().add(PianoIdJsonAdapterFactory()).build()
    val prefsStorage: PrefsStorage = PrefsStorage(appContext, moshi)
    val okHttpClient: OkHttpClient = OkHttpClient.Builder()
        .addInterceptor(
            HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY),
        )
        .build()

    companion object {
        @JvmStatic
        private var instance: SimpleDependenciesProvider? = null

        @JvmStatic
        fun getInstance(context: Context): SimpleDependenciesProvider {
            if (instance == null) {
                synchronized(SimpleDependenciesProvider::class.java) {
                    if (instance == null) {
                        instance = SimpleDependenciesProvider(context)
                    }
                }
            }
            return instance!!
        }
    }
}
